### PR TITLE
[ZEN-836] Add option to wait until callback drop in undeclare

### LINKED ISSUE #1818: Ensure that callbacks are never called when subscriber/queryable go out of scope.
### Describe the bug

Ensure that callbacks are never called when subscriber/queryable go out of scope. Also ensure that these callbacks will be dropped before corresponding undeclare call returns.
This is not relevant to rust, but in other binding like C and C++, calling callbacks after subscriber/queryable destruction can result in UB, since there is no borrow checker to ensure that data captured by callbacks is still alive at the moment of callback invocation/destruction.

### To reproduce

It is sufficient to undeclare subscriber/queryable roughly at the same moment when a new message/query arrives.

### System info

any

### PR DESCRIPTION
**Pull request by DenisBiryukov91**
## Description

Add option to wait until related callback execution ends and they are dropped for undeclare() operation for
Publisher, Subscriber, Querier, Queryable, Matching Listener, Transport and Link Event listeners, Sessio and Advanced Publisher, Subscriber and Miss Listener.

### What does this PR do?
Add new api that allows to force undeclare to block until all related callback execution finishes. This is in particular useful for c/c++ (or any other non-rust language) bindings to ensure that once corresponding zenoh entity is destroyed, no callbacks will be invoked and no related resources will be accessed.

### Why is this change needed?
See above.

### Related Issues
Fixes #1818

---
## 🏷️ Label-Based Checklist

Based on the labels applied to this PR, please complete these additional requirements:

**Labels:** `new feature`

## 🆕 New Feature Requirements

Since this PR adds a new feature:

- [x] **Feature scope documented** - Clear description of what the feature does and why it's needed
- [x] **Minimum necessary code** - Implementation is as simple as possible, doesn't overcomplicate the system
- [x] **New APIs well-designed** - Public APIs are intuitive, consistent with existing APIs
- [x] **Comprehensive tests** - All functionality is tested (happy path + edge cases + error cases)
- [x] **Examples provided** - Usage examples in code comments or separate example files
- [x] **Documentation added** - New docs explaining the feature, its use cases, and API
- [x] **Feature flag considered** - Can the feature be enabled/disabled for gradual rollout?
- [x] **Performance impact assessed** - Memory, CPU, storage implications measured
- [x] **Integration tested** - Feature works with existing features

**Consider:** Can this feature be split into smaller, incremental PRs?

**Instructions:**
1. Check off items as you complete them (change `- [ ]` to `- [x]`)
2. The PR checklist CI will verify these are completed

*This checklist updates automatically when labels change, but preserves your checked boxes.*

### CODE REVIEW

File: zenoh/src/api/builders/close.rs [hunk 1]
```diff
@@ -220,10 +222,20 @@ impl<TOutput: Send + 'static> IntoFuture for NolocalJoinHandle<TOutput> {
...
 
 pub(crate) trait Closeable {
     type TClosee: Closee;
     fn get_closee(&self) -> Self::TClosee;
 }
+
+impl CloseBuilder<crate::Session> {
+    /// Block in undeclare operation until all currently running zenoh entities' callbacks (if any) return.
+    #[zenoh_macros::unstable]
+    pub fn wait_until_callback_execution_ends(mut self) -> Self {
```
Comment (OlivierHecart) [inline zenoh/src/api/builders/close.rs]: We should try to find out a shorter name.
What about `wait_callbacks` ?

Comment (DenisBiryukov91) [reply zenoh/src/api/builders/close.rs]: `wait_callbacks_drop` maybe ?

File: zenoh/src/api/cancellation.rs [hunk 2]
```diff
@@ -171,26 +227,32 @@ impl CancellationToken {
...
-                    self.sync_group.close();
-                }
-            }
-        };
+            Some(actions) => Ok(actions.add(on_cancel)),
+            None => Err(on_cancel),
+        }
     }
 
     #[zenoh_macros::internal]
```
Comment (OlivierHecart) [inline zenoh/src/api/cancellation.rs]: No `#[zenoh_macros::pub_visibility_if_internal]` ?

**Review by OlivierHecart — COMMENTED** [review]
See comments

File: zenoh/src/api/cancellation.rs (hunk 2 above)
Comment (DenisBiryukov91) [reply zenoh/src/api/cancellation.rs]: Fixed

File: zenoh/src/api/builders/close.rs (hunk 1 above)
Comment (DenisBiryukov91) [reply zenoh/src/api/builders/close.rs]: Fixed
